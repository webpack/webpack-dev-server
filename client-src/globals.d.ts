declare interface CommunicationClient {
  onOpen(fn: (...args: any[]) => void): void;
  onClose(fn: (...args: any[]) => void): void;
  onMessage(fn: (...args: any[]) => void): void;
}

declare interface CommunicationClientConstructor {
  new (url: string): CommunicationClient; // Defines a constructor that takes a string and returns a GreeterInstance
}

declare const __webpack_dev_server_client__:
  | CommunicationClientConstructor
  | { default: CommunicationClientConstructor }
  | undefined;

declare module "ansi-html-community" {
  function ansiHtmlCommunity(str: string): string;

  namespace ansiHtmlCommunity {
    function setColors(colors: Record<string, string | string[]>): void;
  }

  export = ansiHtmlCommunity;
}
