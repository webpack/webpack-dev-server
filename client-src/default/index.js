'use strict';

/* global __resourceQuery WorkerGlobalScope self */
/* eslint prefer-destructuring: off */

const url = require('url');
const stripAnsi = require('strip-ansi');
const log = require('loglevel').getLogger('webpack-dev-server');
const socket = require('./socket');
const overlay = require('./overlay');

function getCurrentScriptSource() {
  // `document.currentScript` is the most accurate way to find the current script,
  // but is not supported in all browsers.
  if (document.currentScript) { return document.currentScript.getAttribute('src'); }
  // Fall back to getting all scripts in the document.
  const scriptElements = document.scripts || [];
  const currentScript = scriptElements[scriptElements.length - 1];
  if (currentScript) { return currentScript.getAttribute('src'); }
  // Fail as there was no script to use.
  throw new Error('[WDS] Failed to get current script source.');
}

let urlParts;
let hotReload = true;
if (typeof window !== 'undefined') {
  const qs = window.location.search.toLowerCase();
  hotReload = qs.indexOf('hotreload=false') === -1;
}
if (typeof __resourceQuery === 'string' && __resourceQuery) {
  // If this bundle is inlined, use the resource query to get the correct url.
  urlParts = url.parse(__resourceQuery.substr(1));
} else {
  // Else, get the url from the <script> this file was called with.
  let scriptHost = getCurrentScriptSource();
  // eslint-disable-next-line no-useless-escape
  scriptHost = scriptHost.replace(/\/[^\/]+$/, '');
  urlParts = url.parse((scriptHost || '/'), false, true);
}

if (!urlParts.port || urlParts.port === '0') {
  urlParts.port = self.location.port;
}

let hot = false;
let initial = true;
let currentHash = '';
let useWarningOverlay = false;
let useErrorOverlay = false;
let useProgress = false;

const INFO = 'info';
const WARNING = 'warning';
const ERROR = 'error';
const NONE = 'none';

// Set the default log level
log.setDefaultLevel(INFO);

// Send messages to the outside, so plugins can consume it.
function sendMsg(type, data) {
  if (
    typeof self !== 'undefined' &&
  (typeof WorkerGlobalScope === 'undefined' ||
  !(self instanceof WorkerGlobalScope))
  ) {
    self.postMessage({
      type: `webpack${type}`,
      data
    }, '*');
  }
}

const onSocketMsg = {
  hot() {
    hot = true;
    log.info('[WDS] Hot Module Replacement enabled.');
  },
  invalid() {
    log.info('[WDS] App updated. Recompiling...');
    // fixes #1042. overlay doesn't clear if errors are fixed but warnings remain.
    if (useWarningOverlay || useErrorOverlay) overlay.clear();
    sendMsg('Invalid');
  },
  hash(hash) {
    currentHash = hash;
  },
  'still-ok': function stillOk() {
    log.info('[WDS] Nothing changed.');
    if (useWarningOverlay || useErrorOverlay) overlay.clear();
    sendMsg('StillOk');
  },
  'log-level': function logLevel(level) {
    const hotCtx = require.context('webpack/hot', false, /^\.\/log$/);
    if (hotCtx.keys().indexOf('./log') !== -1) {
      hotCtx('./log').setLogLevel(level);
    }
    switch (level) {
      case INFO:
      case ERROR:
        log.setLevel(level);
        break;
      case WARNING:
        // loglevel's warning name is different from webpack's
        log.setLevel('warn');
        break;
      case NONE:
        log.disableAll();
        break;
      default:
        log.error(`[WDS] Unknown clientLogLevel '${level}'`);
    }
  },
  overlay(value) {
    if (typeof document !== 'undefined') {
      if (typeof (value) === 'boolean') {
        useWarningOverlay = false;
        useErrorOverlay = value;
      } else if (value) {
        useWarningOverlay = value.warnings;
        useErrorOverlay = value.errors;
      }
    }
  },
  progress(progress) {
    if (typeof document !== 'undefined') {
      useProgress = progress;
    }
  },
  'progress-update': function progressUpdate(data) {
    if (useProgress) log.info(`[WDS] ${data.percent}% - ${data.msg}.`);
  },
  ok() {
    sendMsg('Ok');
    if (useWarningOverlay || useErrorOverlay) overlay.clear();
    if (initial) return initial = false; // eslint-disable-line no-return-assign
    reloadApp();
  },
  'content-changed': function contentChanged() {
    log.info('[WDS] Content base changed. Reloading...');
    self.location.reload();
  },
  warnings(warnings) {
    log.warn('[WDS] Warnings while compiling.');
    const strippedWarnings = warnings.map(warning => stripAnsi(warning));
    sendMsg('Warnings', strippedWarnings);
    for (let i = 0; i < strippedWarnings.length; i++) { log.warn(strippedWarnings[i]); }
    if (useWarningOverlay) overlay.showMessage(warnings);

    if (initial) return initial = false; // eslint-disable-line no-return-assign
    reloadApp();
  },
  errors(errors) {
    log.error('[WDS] Errors while compiling. Reload prevented.');
    const strippedErrors = errors.map(error => stripAnsi(error));
    sendMsg('Errors', strippedErrors);
    for (let i = 0; i < strippedErrors.length; i++) { log.error(strippedErrors[i]); }
    if (useErrorOverlay) overlay.showMessage(errors);
    initial = false;
  },
  error(error) {
    log.error(error);
  },
  close() {
    log.error('[WDS] Disconnected!');
    sendMsg('Close');
  }
};

let hostname = urlParts.hostname;
let protocol = urlParts.protocol;


// check ipv4 and ipv6 `all hostname`
if (hostname === '0.0.0.0' || hostname === '::') {
  // why do we need this check?
  // hostname n/a for file protocol (example, when using electron, ionic)
  // see: https://github.com/webpack/webpack-dev-server/pull/384
  // eslint-disable-next-line no-bitwise
  if (self.location.hostname && !!~self.location.protocol.indexOf('http')) {
    hostname = self.location.hostname;
  }
}

// `hostname` can be empty when the script path is relative. In that case, specifying
// a protocol would result in an invalid URL.
// When https is used in the app, secure websockets are always necessary
// because the browser doesn't accept non-secure websockets.
if (hostname && (self.location.protocol === 'https:' || urlParts.hostname === '0.0.0.0')) {
  protocol = self.location.protocol;
}

const socketUrl = url.format({
  protocol,
  auth: urlParts.auth,
  hostname,
  port: urlParts.port,
  pathname: urlParts.path == null || urlParts.path === '/' ? '/sockjs-node' : urlParts.path
});

socket(socketUrl, onSocketMsg);

let isUnloading = false;
self.addEventListener('beforeunload', () => {
  isUnloading = true;
});

function reloadApp() {
  if (isUnloading || !hotReload) {
    return;
  }
  if (hot) {
    log.info('[WDS] App hot update...');
    // eslint-disable-next-line global-require
    const hotEmitter = require('webpack/hot/emitter');
    hotEmitter.emit('webpackHotUpdate', currentHash);
    if (typeof self !== 'undefined' && self.window) {
      // broadcast update to window
      self.postMessage(`webpackHotUpdate${currentHash}`, '*');
    }
  } else {
    let rootWindow = self;
    // use parent window for reload (in case we're in an iframe with no valid src)
    const intervalId = self.setInterval(() => {
      if (rootWindow.location.protocol !== 'about:') {
        // reload immediately if protocol is valid
        applyReload(rootWindow, intervalId);
      } else {
        rootWindow = rootWindow.parent;
        if (rootWindow.parent === rootWindow) {
          // if parent equals current window we've reached the root which would continue forever, so trigger a reload anyways
          applyReload(rootWindow, intervalId);
        }
      }
    });
  }

  function applyReload(rootWindow, intervalId) {
    clearInterval(intervalId);
    log.info('[WDS] App updated. Reloading...');
    rootWindow.location.reload();
  }
}
