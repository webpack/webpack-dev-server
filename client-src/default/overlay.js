'use strict';

// The error overlay is inspired (and mostly copied) from Create React App (https://github.com/facebookincubator/create-react-app)
// They, in turn, got inspired by webpack-hot-middleware (https://github.com/glenjamin/webpack-hot-middleware).

const ansiHTML = require('ansi-html');
const { AllHtmlEntities } = require('html-entities');

const entities = new AllHtmlEntities();
const colors = {
  reset: ['transparent', 'transparent'],
  black: '181818',
  red: 'E36049',
  green: 'B3CB74',
  yellow: 'FFD080',
  blue: '7CAFC2',
  magenta: '7FACCA',
  cyan: 'C3C2EF',
  lightgrey: 'EBE7E3',
  darkgrey: '6D7891',
};

let overlayIframe = null;
let overlayDiv = null;
let lastOnOverlayDivReady = null;

ansiHTML.setColors(colors);

function createOverlayIframe(onIframeLoad) {
  const iframe = document.createElement('iframe');
  iframe.id = 'webpack-dev-server-client-overlay';
  iframe.src = 'about:blank';
  iframe.style.position = 'fixed';
  iframe.style.left = 0;
  iframe.style.top = 0;
  iframe.style.right = 0;
  iframe.style.bottom = 0;
  iframe.style.width = '100vw';
  iframe.style.height = '100vh';
  iframe.style.border = 'none';
  iframe.style.zIndex = 9999999999;
  iframe.onload = onIframeLoad;
  return iframe;
}

function addOverlayDivTo(iframe) {
  const div = iframe.contentDocument.createElement('div');
  div.id = 'webpack-dev-server-client-overlay-div';
  div.style.position = 'fixed';
  div.style.boxSizing = 'border-box';
  div.style.left = 0;
  div.style.top = 0;
  div.style.right = 0;
  div.style.bottom = 0;
  div.style.width = '100vw';
  div.style.height = '100vh';
  div.style.backgroundColor = 'rgba(0, 0, 0, 0.85)';
  div.style.color = '#E8E8E8';
  div.style.fontFamily = 'Menlo, Consolas, monospace';
  div.style.fontSize = 'large';
  div.style.padding = '2rem';
  div.style.lineHeight = '1.2';
  div.style.whiteSpace = 'pre-wrap';
  div.style.overflow = 'auto';
  iframe.contentDocument.body.appendChild(div);
  return div;
}

function ensureOverlayDivExists(onOverlayDivReady) {
  if (overlayDiv) {
    // Everything is ready, call the callback right away.
    onOverlayDivReady(overlayDiv);
    return;
  }

  // Creating an iframe may be asynchronous so we'll schedule the callback.
  // In case of multiple calls, last callback wins.
  lastOnOverlayDivReady = onOverlayDivReady;

  if (overlayIframe) {
    // We've already created it.
    return;
  }

  // Create iframe and, when it is ready, a div inside it.
  overlayIframe = createOverlayIframe(() => {
    overlayDiv = addOverlayDivTo(overlayIframe);
    // Now we can talk!
    lastOnOverlayDivReady(overlayDiv);
  });

  // Zalgo alert: onIframeLoad() will be called either synchronously
  // or asynchronously depending on the browser.
  // We delay adding it so `overlayIframe` is set when `onIframeLoad` fires.
  document.body.appendChild(overlayIframe);
}

// Successful compilation.
function clear() {
  if (!overlayDiv) {
    // It is not there in the first place.
    return;
  }

  // Clean up and reset internal state.
  document.body.removeChild(overlayIframe);
  overlayDiv = null;
  overlayIframe = null;
  lastOnOverlayDivReady = null;
}

// Compilation with errors (e.g. syntax error or missing modules).
function showMessage(messages) {
  ensureOverlayDivExists((div) => {
    // Make it look similar to our terminal.
    const errorMessage = messages[0].message || messages[0];
    const text = ansiHTML(entities.encode(errorMessage));

    div.innerHTML = `<span style="color: #${colors.red}">Failed to compile.</span><br><br>${text}`;
  });
}

module.exports = {
  clear,
  showMessage,
};
