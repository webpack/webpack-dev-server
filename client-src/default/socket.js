'use strict';

/* global __webpack_dev_server_client__ */
/* eslint-disable
  camelcase
*/

// this SockJSClient is here as a default fallback, in case inline mode
// is off or the client is not injected. This will be switched to
// WebsocketClient when it becomes the default

// important: the path to SockJSClient here is made to work in the 'client'
// directory, but is updated via the webpack compilation when compiled from
// the 'client-src' directory
const Client =
  typeof __webpack_dev_server_client__ !== 'undefined'
    ? __webpack_dev_server_client__
    : // eslint-disable-next-line import/no-unresolved
      require('./clients/SockJSClient');

let retries = 0;
let client = null;

const socket = function initSocket(url, handlers) {
  client = new Client(url);

  client.onOpen(() => {
    retries = 0;
  });

  client.onClose(() => {
    if (retries === 0) {
      handlers.close();
    }

    // Try to reconnect.
    client = null;

    // After 10 retries stop trying, to prevent logspam.
    if (retries <= 10) {
      // Exponentially increase timeout to reconnect.
      // Respectfully copied from the package `got`.
      // eslint-disable-next-line no-mixed-operators, no-restricted-properties
      const retryInMs = 1000 * Math.pow(2, retries) + Math.random() * 100;
      retries += 1;

      setTimeout(() => {
        socket(url, handlers);
      }, retryInMs);
    }
  });

  client.onMessage((data) => {
    const msg = JSON.parse(data);
    if (handlers[msg.type]) {
      handlers[msg.type](msg.data);
    }
  });
};

module.exports = socket;
