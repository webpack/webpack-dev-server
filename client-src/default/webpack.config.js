'use strict';

module.exports = {
  mode: 'production',
  module: {
    rules: [
      {
        test: /\.js$/,
        exclude: /node_modules|web_modules/,
        use: [
          {
            loader: 'babel-loader'
          }
        ]
      }
    ]
  }
};
