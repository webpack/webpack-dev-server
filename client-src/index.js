/* global __resourceQuery, __webpack_hash__ */
/// <reference types="webpack/module" />
import webpackHotLog from "webpack/hot/log.js";
import stripAnsi from "./utils/stripAnsi.js";
import parseURL from "./utils/parseURL.js";
import socket from "./socket.js";
import { formatProblem, createOverlay } from "./overlay.js";
import { log, logEnabledFeatures, setLogLevel } from "./utils/log.js";
import sendMessage from "./utils/sendMessage.js";
import reloadApp from "./utils/reloadApp.js";
import createSocketURL from "./utils/createSocketURL.js";

/**
 * @typedef {Object} Options
 * @property {boolean} hot
 * @property {boolean} liveReload
 * @property {boolean} progress
 * @property {boolean | { warnings?: boolean, errors?: boolean, runtimeErrors?: boolean, trustedTypesPolicyName?: string }} overlay
 * @property {string} [logging]
 * @property {number} [reconnect]
 */

/**
 * @typedef {Object} Status
 * @property {boolean} isUnloading
 * @property {string} currentHash
 * @property {string} [previousHash]
 */

/**
 * @type {Status}
 */
const status = {
  isUnloading: false,
  // TODO Workaround for webpack v4, `__webpack_hash__` is not replaced without HotModuleReplacement
  // eslint-disable-next-line camelcase
  currentHash: typeof __webpack_hash__ !== "undefined" ? __webpack_hash__ : "",
};

/** @type {Options} */
const options = {
  hot: false,
  liveReload: false,
  progress: false,
  overlay: false,
};
const parsedResourceQuery = parseURL(__resourceQuery);

const enabledFeatures = {
  "Hot Module Replacement": false,
  "Live Reloading": false,
  Progress: false,
  Overlay: false,
};

if (parsedResourceQuery.hot === "true") {
  options.hot = true;
  enabledFeatures["Hot Module Replacement"] = true;
}

if (parsedResourceQuery["live-reload"] === "true") {
  options.liveReload = true;
  enabledFeatures["Live Reloading"] = true;
}

if (parsedResourceQuery.progress === "true") {
  options.progress = true;
  enabledFeatures.Progress = true;
}

if (parsedResourceQuery.overlay) {
  try {
    options.overlay = JSON.parse(parsedResourceQuery.overlay);
  } catch (e) {
    log.error("Error parsing overlay options from resource query:", e);
  }

  // Fill in default "true" params for partially-specified objects.
  if (typeof options.overlay === "object") {
    options.overlay = {
      errors: true,
      warnings: true,
      runtimeErrors: true,
      ...options.overlay,
    };
  }
  enabledFeatures.Overlay = true;
}

if (parsedResourceQuery.logging) {
  options.logging = parsedResourceQuery.logging;
}

if (typeof parsedResourceQuery.reconnect !== "undefined") {
  options.reconnect = Number(parsedResourceQuery.reconnect);
}

/**
 * @param {string} level
 */
function setAllLogLevel(level) {
  // This is needed because the HMR logger operate separately from dev server logger
  webpackHotLog.setLogLevel(
    level === "verbose" || level === "log" ? "info" : level
  );
  setLogLevel(level);
}

if (options.logging) {
  setAllLogLevel(options.logging);
}

logEnabledFeatures(enabledFeatures);

self.addEventListener("beforeunload", () => {
  status.isUnloading = true;
});

const overlay =
  typeof window !== "undefined"
    ? createOverlay(
        typeof options.overlay === "object"
          ? {
              trustedTypesPolicyName: options.overlay.trustedTypesPolicyName,
              catchRuntimeError: options.overlay.runtimeErrors,
            }
          : {
              trustedTypesPolicyName: false,
              catchRuntimeError: options.overlay,
            }
      )
    : { send: () => {} };

const onSocketMessage = {
  hot() {
    if (parsedResourceQuery.hot === "false") {
      return;
    }

    options.hot = true;
  },
  liveReload() {
    if (parsedResourceQuery["live-reload"] === "false") {
      return;
    }

    options.liveReload = true;
  },
  invalid() {
    log.info("App updated. Recompiling...");

    // Fixes #1042. overlay doesn't clear if errors are fixed but warnings remain.
    if (options.overlay) {
      overlay.send({ type: "DISMISS" });
    }

    sendMessage("Invalid");
  },
  /**
   * @param {string} hash
   */
  hash(hash) {
    status.previousHash = status.currentHash;
    status.currentHash = hash;
  },
  logging: setAllLogLevel,
  /**
   * @param {boolean} value
   */
  overlay(value) {
    if (typeof document === "undefined") {
      return;
    }

    options.overlay = value;
  },
  /**
   * @param {number} value
   */
  reconnect(value) {
    if (parsedResourceQuery.reconnect === "false") {
      return;
    }

    options.reconnect = value;
  },
  /**
   * @param {boolean} value
   */
  progress(value) {
    options.progress = value;
  },
  /**
   * @param {{ pluginName?: string, percent: number, msg: string }} data
   */
  "progress-update": function progressUpdate(data) {
    if (options.progress) {
      log.info(
        `${data.pluginName ? `[${data.pluginName}] ` : ""}${data.percent}% - ${
          data.msg
        }.`
      );
    }

    sendMessage("Progress", data);
  },
  "still-ok": function stillOk() {
    log.info("Nothing changed.");

    if (options.overlay) {
      overlay.send({ type: "DISMISS" });
    }

    sendMessage("StillOk");
  },
  ok() {
    sendMessage("Ok");

    if (options.overlay) {
      overlay.send({ type: "DISMISS" });
    }

    reloadApp(options, status);
  },
  // TODO: remove in v5 in favor of 'static-changed'
  /**
   * @param {string} file
   */
  "content-changed": function contentChanged(file) {
    log.info(
      `${
        file ? `"${file}"` : "Content"
      } from static directory was changed. Reloading...`
    );

    self.location.reload();
  },
  /**
   * @param {string} file
   */
  "static-changed": function staticChanged(file) {
    log.info(
      `${
        file ? `"${file}"` : "Content"
      } from static directory was changed. Reloading...`
    );

    self.location.reload();
  },
  /**
   * @param {Error[]} warnings
   * @param {any} params
   */
  warnings(warnings, params) {
    log.warn("Warnings while compiling.");

    const printableWarnings = warnings.map((error) => {
      const { header, body } = formatProblem("warning", error);

      return `${header}\n${stripAnsi(body)}`;
    });

    sendMessage("Warnings", printableWarnings);

    for (let i = 0; i < printableWarnings.length; i++) {
      log.warn(printableWarnings[i]);
    }

    const needShowOverlayForWarnings =
      typeof options.overlay === "boolean"
        ? options.overlay
        : options.overlay && options.overlay.warnings;

    if (needShowOverlayForWarnings) {
      overlay.send({
        type: "BUILD_ERROR",
        level: "warning",
        messages: warnings,
      });
    }

    if (params && params.preventReloading) {
      return;
    }

    reloadApp(options, status);
  },
  /**
   * @param {Error[]} errors
   */
  errors(errors) {
    log.error("Errors while compiling. Reload prevented.");

    const printableErrors = errors.map((error) => {
      const { header, body } = formatProblem("error", error);

      return `${header}\n${stripAnsi(body)}`;
    });

    sendMessage("Errors", printableErrors);

    for (let i = 0; i < printableErrors.length; i++) {
      log.error(printableErrors[i]);
    }

    const needShowOverlayForErrors =
      typeof options.overlay === "boolean"
        ? options.overlay
        : options.overlay && options.overlay.errors;

    if (needShowOverlayForErrors) {
      overlay.send({
        type: "BUILD_ERROR",
        level: "error",
        messages: errors,
      });
    }
  },
  /**
   * @param {Error} error
   */
  error(error) {
    log.error(error);
  },
  close() {
    log.info("Disconnected!");

    if (options.overlay) {
      overlay.send({ type: "DISMISS" });
    }

    sendMessage("Close");
  },
};

const socketURL = createSocketURL(parsedResourceQuery);

socket(socketURL, onSocketMessage, options.reconnect);
