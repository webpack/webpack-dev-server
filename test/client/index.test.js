'use strict';

/* global self */

describe('index', () => {
  let log;
  let socket;
  let overlay;
  let reloadApp;
  let sendMessage;
  let onSocketMessage;
  const locationValue = self.location;
  const resourceQueryValue = global.__resourceQuery;

  beforeEach(() => {
    global.__resourceQuery = 'foo';

    // log
    jest.setMock('../../client-src/default/utils/log.js', {
      log: {
        info: jest.fn(),
        warn: jest.fn(),
        error: jest.fn(),
      },
    });
    log = require('../../client-src/default/utils/log');

    // socket
    jest.setMock('../../client-src/default/socket.js', jest.fn());
    socket = require('../../client-src/default/socket');

    // overlay
    jest.setMock('../../client-src/default/overlay.js', {
      clear: jest.fn(),
      showMessage: jest.fn(),
    });
    overlay = require('../../client-src/default/overlay');

    // reloadApp
    jest.setMock('../../client-src/default/utils/reloadApp.js', jest.fn());
    reloadApp = require('../../client-src/default/utils/reloadApp');

    // sendMessage
    jest.setMock('../../client-src/default/utils/sendMessage.js', jest.fn());
    sendMessage = require('../../client-src/default/utils/sendMessage');

    // createSocketUrl
    jest.setMock(
      '../../client-src/default/utils/createSocketUrl.js',
      () => 'mock-url'
    );

    // issue: https://github.com/jsdom/jsdom/issues/2112
    delete window.location;
    window.location = {
      ...locationValue,
      reload: jest.fn(),
    };

    require('../../client-src/default');
    onSocketMessage = socket.mock.calls[0][1];
  });

  afterEach(() => {
    global.__resourceQuery = resourceQueryValue;
    Object.assign(self, locationValue);
    jest.resetAllMocks();
    jest.resetModules();
  });

  test('should set arguments into socket function', () => {
    expect(socket.mock.calls[0]).toMatchSnapshot();
  });

  test('should run onSocketMessage.hot', () => {
    onSocketMessage.hot();
    expect(log.log.info.mock.calls[0][0]).toMatchSnapshot();
  });

  test('should run onSocketMessage.liveReload', () => {
    onSocketMessage.liveReload();
    expect(log.log.info.mock.calls[0][0]).toMatchSnapshot();
  });

  test('should run onSocketMessage.invalid', () => {
    onSocketMessage.invalid();
    expect(log.log.info.mock.calls[0][0]).toMatchSnapshot();
    expect(sendMessage.mock.calls[0][0]).toMatchSnapshot();
    expect(overlay.clear).not.toBeCalled();

    // change flags
    onSocketMessage.overlay(true);
    onSocketMessage.invalid();
    expect(overlay.clear).toBeCalled();
  });

  test("should run onSocketMessage['still-ok']", () => {
    onSocketMessage['still-ok']();
    expect(log.log.info.mock.calls[0][0]).toMatchSnapshot();
    expect(sendMessage.mock.calls[0][0]).toMatchSnapshot();
    expect(overlay.clear).not.toBeCalled();

    // change flags
    onSocketMessage.overlay(true);
    onSocketMessage['still-ok']();
    expect(overlay.clear).toBeCalled();
  });

  // TODO: need to mock require.context
  test.skip("should run onSocketMessage['logging']", () => {
    onSocketMessage.logging();
  });

  test("should run onSocketMessage.progress and onSocketMessage['progress-update']", () => {
    onSocketMessage.progress(false);
    onSocketMessage['progress-update']({
      msg: 'mock-msg',
      percent: '12',
    });
    expect(log.log.info).not.toBeCalled();
    expect(sendMessage.mock.calls[0][0]).toMatchSnapshot();

    onSocketMessage.progress(true);
    onSocketMessage['progress-update']({
      msg: 'mock-msg',
      percent: '12',
    });
    expect(log.log.info.mock.calls[0][0]).toMatchSnapshot();
  });

  test('should run onSocketMessage.overlay with an argument is Object', () => {
    onSocketMessage.overlay({
      warnings: true,
      errors: true,
    });

    // check if flags have changed
    onSocketMessage.invalid();
    expect(overlay.clear).toBeCalled();
  });

  test('should run onSocketMessage.ok', () => {
    {
      const res = onSocketMessage.ok();
      expect(sendMessage.mock.calls[0][0]).toMatchSnapshot();
      expect(res).toEqual(false);
    }

    // change flags
    {
      onSocketMessage.errors([]);
      onSocketMessage.hash('mock-hash');

      const res = onSocketMessage.ok();
      expect(reloadApp).toBeCalled();
      expect(reloadApp.mock.calls[0][0]).toMatchSnapshot();
      // eslint-disable-next-line no-undefined
      expect(res).toEqual(undefined);
    }
  });

  test("should run onSocketMessage['content-changed']", () => {
    onSocketMessage['content-changed']();

    expect(log.log.info.mock.calls[0][0]).toMatchSnapshot();
    expect(self.location.reload).toBeCalled();
  });

  test('should run onSocketMessage.warnings', () => {
    {
      const res = onSocketMessage.warnings([
        'warn1',
        '\u001B[4mwarn2\u001B[0m',
        'warn3',
      ]);
      expect(res).toEqual(false);

      expect(log.log.warn.mock.calls[0][0]).toMatchSnapshot();
      expect(sendMessage.mock.calls[0][0]).toMatchSnapshot();
      expect(log.log.warn.mock.calls.splice(1)).toMatchSnapshot();
    }

    // change flags
    {
      onSocketMessage.overlay({
        warnings: true,
      });
      const res = onSocketMessage.warnings([]);
      // eslint-disable-next-line no-undefined
      expect(res).toEqual(undefined);

      expect(overlay.showMessage).toBeCalled();
      expect(reloadApp).toBeCalled();
    }
  });

  test('should run onSocketMessage.errors', () => {
    onSocketMessage.errors(['error1', '\u001B[4error2\u001B[0m', 'error3']);

    expect(log.log.error.mock.calls[0][0]).toMatchSnapshot();
    expect(sendMessage.mock.calls[0][0]).toMatchSnapshot();
    expect(log.log.error.mock.calls.splice(1)).toMatchSnapshot();

    // change flags
    onSocketMessage.overlay({
      errors: true,
    });
    onSocketMessage.errors([]);
    expect(overlay.showMessage).toBeCalled();
  });

  test('should run onSocketMessage.error', () => {
    onSocketMessage.error('error!!');
    expect(log.log.error.mock.calls[0][0]).toMatchSnapshot();
  });

  test('should run onSocketMessage.close', () => {
    onSocketMessage.close();
    expect(log.log.error.mock.calls[0][0]).toMatchSnapshot();
    expect(sendMessage.mock.calls[0][0]).toMatchSnapshot();
  });

  test('should run onSocketMessage.close (hot enabled)', () => {
    // enabling hot
    onSocketMessage.hot();
    onSocketMessage.close();
    expect(log.log.error.mock.calls[0][0]).toMatchSnapshot();
    expect(sendMessage.mock.calls[0][0]).toMatchSnapshot();
  });

  test('should run onSocketMessage.close (liveReload enabled)', () => {
    // enabling liveReload
    onSocketMessage.liveReload();
    onSocketMessage.close();
    expect(log.log.error.mock.calls[0][0]).toMatchSnapshot();
    expect(sendMessage.mock.calls[0][0]).toMatchSnapshot();
  });
});
