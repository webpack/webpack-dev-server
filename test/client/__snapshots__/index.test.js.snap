// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index should run onSocketMessage.close 1`] = `"[WDS] Disconnected!"`;

exports[`index should run onSocketMessage.close 2`] = `"Close"`;

exports[`index should run onSocketMessage.error 1`] = `"error!!"`;

exports[`index should run onSocketMessage.errors 1`] = `"[WDS] Errors while compiling. Reload prevented."`;

exports[`index should run onSocketMessage.errors 2`] = `"Errors"`;

exports[`index should run onSocketMessage.errors 3`] = `
Array [
  Array [
    "error1",
  ],
  Array [
    "error2",
  ],
  Array [
    "error3",
  ],
]
`;

exports[`index should run onSocketMessage.hot 1`] = `"[WDS] Hot Module Replacement enabled."`;

exports[`index should run onSocketMessage.invalid 1`] = `"[WDS] App updated. Recompiling..."`;

exports[`index should run onSocketMessage.invalid 2`] = `"Invalid"`;

exports[`index should run onSocketMessage.liveReload 1`] = `"[WDS] Live Reloading enabled."`;

exports[`index should run onSocketMessage.ok 1`] = `"Ok"`;

exports[`index should run onSocketMessage.ok 2`] = `
Object {
  "hot": false,
  "hotReload": true,
  "initial": false,
  "liveReload": false,
  "useErrorOverlay": false,
  "useProgress": false,
  "useWarningOverlay": false,
}
`;

exports[`index should run onSocketMessage.progress and onSocketMessage['progress-update'] 1`] = `"Progress"`;

exports[`index should run onSocketMessage.progress and onSocketMessage['progress-update'] 2`] = `"[WDS] 12% - mock-msg."`;

exports[`index should run onSocketMessage.warnings 1`] = `"[WDS] Warnings while compiling."`;

exports[`index should run onSocketMessage.warnings 2`] = `"Warnings"`;

exports[`index should run onSocketMessage.warnings 3`] = `
Array [
  Array [
    "warn1",
  ],
  Array [
    "warn2",
  ],
  Array [
    "warn3",
  ],
]
`;

exports[`index should run onSocketMessage['content-changed'] 1`] = `"[WDS] Content base changed. Reloading..."`;

exports[`index should run onSocketMessage['still-ok'] 1`] = `"[WDS] Nothing changed."`;

exports[`index should run onSocketMessage['still-ok'] 2`] = `"StillOk"`;

exports[`index should set arguments into socket function 1`] = `
Array [
  "mock-url",
  Object {
    "close": [Function],
    "content-changed": [Function],
    "error": [Function],
    "errors": [Function],
    "hash": [Function],
    "hot": [Function],
    "invalid": [Function],
    "liveReload": [Function],
    "log-level": [Function],
    "ok": [Function],
    "overlay": [Function],
    "progress": [Function],
    "progress-update": [Function],
    "still-ok": [Function],
    "warnings": [Function],
  },
]
`;
